### ISSUE TITLE
请教Pretrian中断后继续训练的疑问

### ISSUE THREAD

**Issue by VictorSun1996** [jingyaogong/minimind#82]
在执行Pretrain.py时如果没有训练完就中断了
想要在中断前的checkpoint的基础上继续训练，本项目的代码请问应该修改哪个部分呀。
谢谢各位大佬

**Comment by jingyaogong** [comment]
可以看一下你 `./out/pretrain_xx_xx.pth` 的修改时间，就是你中断之前被保存的最后一次checkpoint时间

所以把 `1-pretrain.py` 第105行换成
```
def init_model():
    def count_parameters(model):
        return sum(p.numel() for p in model.parameters() if p.requires_grad)

    tokenizer = AutoTokenizer.from_pretrained('./model/minimind_tokenizer')

    model = Transformer(lm_config).to(args.device)
    moe_path = '_moe' if lm_config.use_moe else ''
    ckp = f'./out/pretrain_{lm_config.dim}{moe_path}.pth'

    state_dict = torch.load(ckp, map_location=args.device)

    # 处理不需要的前缀
    unwanted_prefix = '_orig_mod.'
    for k, v in list(state_dict.items()):
        if k.startswith(unwanted_prefix):
            state_dict[k[len(unwanted_prefix):]] = state_dict.pop(k)

    for k, v in list(state_dict.items()):
        if 'mask' in k:
            del state_dict[k]

    # 加载到模型中
    model.load_state_dict(state_dict, strict=False)
    Logger(f'LLM总参数量：{count_parameters(model) / 1e6:.3f} 百万')
    return model, tokenizer
```
加载它即可

**Comment by VictorSun1996** [comment]
> 可以看一下你 `./out/pretrain_xx_xx.pth` 的修改时间，就是你中断之前被保存的最后一次checkpoint时间
> 
> 所以把 `1-pretrain.py` 第105行换成
> 
> ```
> def init_model():
>     def count_parameters(model):
>         return sum(p.numel() for p in model.parameters() if p.requires_grad)
> 
>     tokenizer = AutoTokenizer.from_pretrained('./model/minimind_tokenizer')
> 
>     model = Transformer(lm_config).to(args.device)
>     moe_path = '_moe' if lm_config.use_moe else ''
>     ckp = f'./out/pretrain_{lm_config.dim}{moe_path}.pth'
> 
>     state_dict = torch.load(ckp, map_location=args.device)
> 
>     # 处理不需要的前缀
>     unwanted_prefix = '_orig_mod.'
>     for k, v in list(state_dict.items()):
>         if k.startswith(unwanted_prefix):
>             state_dict[k[len(unwanted_prefix):]] = state_dict.pop(k)
> 
>     for k, v in list(state_dict.items()):
>         if 'mask' in k:
>             del state_dict[k]
> 
>     # 加载到模型中
>     model.load_state_dict(state_dict, strict=False)
>     Logger(f'LLM总参数量：{count_parameters(model) / 1e6:.3f} 百万')
>     return model, tokenizer
> ```
> 
> 加载它即可

谢谢，明白了。

**Comment by zxywww** [comment]
# 处理不需要的前缀
    unwanted_prefix = '_orig_mod.'
    for k, v in list(state_dict.items()):
        if k.startswith(unwanted_prefix):
            state_dict[k[len(unwanted_prefix):]] = state_dict.pop(k)

    for k, v in list(state_dict.items()):
        if 'mask' in k:
            del state_dict[k]

这段代码在干嘛? 有什么用嘛? 哪来的前缀要去除?
